### DISCUSSION TITLE
Setup a `hybrid_property` returning a list of values from a relationship

### DISCUSSION THREAD

**Discussion by glatterf42** [sqlalchemy/sqlalchemy#12024] [Usage Questions] [answered]
Hi everyone and thanks for maintaining this project, including the extensive documentation! Unfortunately, I'm struggling to implement a data model despite all the documentation and existing discussions, so I would greatly appreciate your guidance on this. 

Essentially, I'm trying to set up a database model for storing data from mathematical optimization software (e.g. [linopy](https://linopy.readthedocs.io/en/latest/)). The first item is called `Indexset`, which mainly has a `.name` and a list of values (`str`/`float`/`int`), which has a differing number of entries for all `Indexset`s. 
In trying to normalize this model, I split the actual data from some meta information, and linked both via a one (`Indexset`)-to-many (`data`) relation. For ease of later use, I want to provide a `.data` attribute on `Indexset` that returns the list of values from the data table. So in code, my setup is:

```python
from typing import Annotated, Literal

import numpy as np

from sqlalchemy import ForeignKey, UniqueConstraint
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import Mapped as Mapped
from sqlalchemy.orm import mapped_column, relationship
from sqlalchemy.types import String

from ixmp4 import db
from ixmp4.data import types

from .. import base


class IndexSet(base.BaseModel):
    ...

    _data_type: Mapped[Literal["float", "int", "str"] | None]

    _data: Mapped[list["IndexSetData"]] = relationship(
        back_populates="indexset"
    )

    @hybrid_property
    def data(self) -> list[float | int | str]:
        return (
            []
            if self._data_type is None
            else np.array([d.value for d in self._data], dtype=self._data_type).tolist()
        )

    # NOTE Without this, setting and retrieving did not work in my facade layer
    @data.inplace.setter
    def _data_setter(self, value: list[float | int | str]) -> None:
        return None


class IndexSetData(base.RootBaseModel):
    indexset: Mapped["IndexSet"] = relationship(back_populates="_data")
    indexset__id: Mapped[Annotated[
    int,
    Column(Integer, ForeignKey("optimization_indexset.id"), nullable=False, index=True),
    ]]
    value: Mapped[str] = mapped_column(String, nullable=False)

    __table_args__ = (UniqueConstraint("indexset__id", "value"),)
``` 

Here, the base classes set up some things like that every table has an `id` primary key and that the `Indexset` table is called `optimization_indexset` behind the scenes. 

This setup works in that I can set and get `.data` as expected when working with instances. However, when I try to select the whole `Indexset` table, it has neither a `_data` nor `data` column. And if I try to select `Indexset.data` explicitly, I receive

```python
>       raise NotImplementedError(str(op))
E       NotImplementedError: <built-in function getitem>

.venv/lib/python3.12/site-packages/sqlalchemy/sql/operators.py:364: NotImplementedError
```

I have tried to mitigate this by introducing an `expression` decorator, but I'm not sure how to properly do that. It would need to return a list of an `Indexset`'s values, potentially of the proper type. 
When I try e.g.

```python
    @data.inplace.expression
    @classmethod
    def _data_expression(cls):
        return (
            db.select(IndexSetData.value.label("data"))
            .where(IndexSetData.indexset__id == cls.id)
            .subquery()
        )
```

When I then try to use it as described [here](https://docs.sqlalchemy.org/en/20/orm/extensions/hybrid.html#join-dependent-relationship-hybrid), I see that this SQL expression is rendered:

```sql
SELECT optimization_indexset.data_type, optimization_indexset.run__id, optimization_indexset.name, optimization_indexset.id, optimization_indexset.created_at, optimization_indexset.created_by, anon_1.data 
FROM optimization_indexset JOIN indexsetdata ON optimization_indexset.id = indexsetdata.indexset__id, (SELECT indexsetdata.value AS data 
FROM indexsetdata, optimization_indexset 
WHERE indexsetdata.indexset__id = optimization_indexset.id) AS anon_1
```

This looks a bit contrived and I would hope we need only one `join` between the two tables. I also see a warning about a cartesian product between `anon_1` and `optimization_indexset`. This code does run, but it doesn't return any rows, despite the test creating two `Indexset`s beforehand. (I know that this works because when I simply select the `Indexset` table and then `Indexset.data` after getting every `Indexset` again, the test passes. In a sense, this is just an attempt to optimize the query.) 

Please let me know if what I'm trying to do is possible at all, of which I'm not too sure. If this should work, please tell me what I'm missing or where I can learn more. 
Thanks in advance :)

**Comment by CaselIT** [comment, marked as answer]
Hi,

First thing you have to ask is if you want to use this data property do query the db, using it inside a select or a where. If not (and that seems to be the case from your example) just use a normal python property.

What you are trying to do is somewhat similar to association proxy https://docs.sqlalchemy.org/en/20/orm/extensions/associationproxy.html#sqlalchemy.ext.associationproxy.association_proxy but I don't think it supports converting the values to int/str/float on the fly

**Reply by glatterf42** [reply]
Thanks for the quick reply! 
Do you mean "query" as in `select(Indexset).where(Indexset.data == ...)`? Because I don't want to do that, as far as I know. However, I would like a `select(Indexset)` to return a `data` column. Does that work with normal python properties?

I'll take a look at association proxies :)

**Reply by CaselIT** [reply]
sure, `select(Indexset)` will return python instances when run in a session, so a python property will work in any other python instance.
You will probably want to eager-load the relationship either by default or using an option in the select.

**Reply by glatterf42** [reply]
Thanks again, but I still can't get it to work with a normal property. I've changed nothing of the model setup except for making `data` a `@property` and removing the `setter`, since the [docs seem to recommend](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#joined-eager-loading) setting `.options(joinedload=Indexset._data)` on the `select` statement. 
So I'm now running this statement:

```python
        _exc = select(IndexSet).options(joinedload(IndexSet._data))
        print(_exc)
        _exc = _exc.order_by(IndexSet.id.asc())

        if self.session.bind is not None:
            with self.engine.connect() as con:
                return pd.read_sql(_exc, con=con).replace([np.nan], [None])
        else:
            raise ProgrammingError("Database session is closed.")
```

Where `pd` is `pandas`, of course, and `ProgrammingError` is an internally managed error. The `print()` statement tells me the SQL expression is the following:

```sql
SELECT optimization_indexset._data_type, optimization_indexset.run__id, optimization_indexset.name, optimization_indexset.id, optimization_indexset.created_at, optimization_indexset.created_by, indexsetdata_1.indexset__id, indexsetdata_1.value, indexsetdata_1.id AS id_1 
FROM optimization_indexset LEFT OUTER JOIN indexsetdata AS indexsetdata_1 ON optimization_indexset.id = indexsetdata_1.indexset__id
```

Which returns the `value` and `id` columns of the `IndexSetData`, so all the data are present with that, but not quite in the desired shape yet: there's still no `data` column in my result. However, the result does produce some rows, which is more than before :)

(Just noticed that all of the following is "using the property inside a `select`", which I said I didn't want to do, so please ignore. Just leaving this in for completeness.)
(When I try to `... = select(IndexSet, IndexSet.data).options(...` instead, I receive this error:

```python
E       sqlalchemy.exc.ArgumentError: Column expression, FROM clause, or other columns clause element expected, got <property object at 0x7622b5312340>.

.venv/lib/python3.12/site-packages/sqlalchemy/sql/coercions.py:518: ArgumentError
```
)
So unfortunately, it looks like I'm also missing something about eager-loading relationships.

**Reply by glatterf42** [reply]
I'm thinking the assumption that I don't want to use `.data` in a `select` might be wrong. For the kind of query I want to run, I think the SQL expression needs to include the `data` column already. One option might be to break this down: load every `Indexset` separately, then store its `.data` as a column in the collecting `pd.DataFrame`. However, this seems to be inefficient, so I would like to avoid it if possible. 

Maybe association proxies are then the way to go. The current query of interest is in the `tabulate()` function, which should only be used to get an overview of what `Indexset`s are available. If you want to perform work on one, you should load it separately, anyway. So for such an overview function, it might be acceptable to display the `data` column as the strings it is stored as. To avoid losing the option to filter for `Indexset`s of a specific type, I might then also display the `_data_type` column, which is otherwise only supposed to help with internal storage.

**Reply by glatterf42** [reply]
But even with association proxies: if I change the `data` attribute to 

```python
data: AssociationProxy[list[str]] = association_proxy("_data", "value")
```

The tests are running as before, i.e. I get some results, but these don't include a `data` column. It does not matter whether I load the relationship lazily or eagerly for this.

So maybe there's no way around figuring out how to convert this into an SQL expression (if that is possible):

```python
    @hybrid_property
    def data(self) -> list[float | int | str]:
        return (
            []
            if self._data_type is None
            else np.array([d.value for d in self._data], dtype=self._data_type).tolist()
        )
```

But please let me know if I'm again missing something with all of this.

**Reply by CaselIT** [reply]
What you are trying to do will not work since you are trying to use orm constructs while not using the orm.
Note that I said "when running things in a session", here you are running the sql directly on a connection, so you will never have orm instances.

If you want to keep reading the data into a pandas dataframe directly, you will need to think at an sql query that can generate the data you are looking for. You probably need to to join `IndexSet` with `IndexSetData`

**Reply by glatterf42** [reply]
Thanks, another good point. 

Rethinking the purpose of the query, I have come to realize that it's supposed to give an overview of which objects are already defined in the DB. Having a variable-length column in the resulting `DataFrame` would probably hinder that -- as well as occur performance cost for loading all `.data` of all `Indexset`s per default. So I'm adding a flag to the function executing the statement for now: if users want to load the data, they can indicate that, but will be made aware of the additional cost. If they only want to load individual data, they can already do that since all other queries use the session.

I'll select your initial reply as the answer, even though we didn't actually do what I thought I wanted to do initially, but your insightful comment sent me down the path of recognizing my needs better. Thanks again :)
